### PR TITLE
UCP/AM: Zcopy protocol for ucp_am_send_nbx

### PR DESCRIPTION
**Pull request by brminich**
## What
Implement zcopy protocol for `ucp_am_send_nbx`

### CODE REVIEW

**Comment by brminich** [comment]
@dmitrygx, @hoopoepg, plz review

**Review by hoopoepg — APPROVED** [review]

File: src/ucp/core/ucp_am.c [hunk 1]
```diff
@@ -551,7 +603,25 @@ ucp_am_send_req(ucp_request_t *req, size_t count,
         max_short = ucp_am_get_short_max(req, msg_config);
     }
 
-    status = ucp_request_send_start(req, max_short, SIZE_MAX, SIZE_MAX,
+    if (user_header_length + sizeof(ucp_am_first_hdr_t) + 1 >
```
Comment (dmitrygx) [inline src/ucp/core/ucp_am.c:606]: need `()` around `user_header_length + sizeof(ucp_am_first_hdr_t) + 1`

**Review by dmitrygx — APPROVED** [review]

File: src/ucp/core/ucp_am.c (hunk 1 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:606]: ok, will apply with other comments

**Comment by brminich** [comment]
@yosefe, plz review

File: src/ucp/core/ucp_am.c [hunk 2]
```diff
@@ -470,43 +479,86 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
...
+    ucp_request_t *req            = ucs_container_of(self, ucp_request_t,
+                                                     send.uct);
+    unsigned user_hdr_length      = req->send.msg_proto.am.header_length;
+    size_t hdr_length             = sizeof(ucp_am_reply_hdr_t) +
+                                    user_hdr_length;
+    ucp_am_reply_hdr_t *reply_hdr = ucs_alloca(hdr_length);
+
+    reply_hdr->ep_ptr             = ucp_request_get_dest_ep_ptr(req);
+
+    return ucp_am_zcopy_single_common(self, (ucp_am_hdr_t*)reply_hdr,
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:526]: &reply_hdr->super

File: src/ucp/core/ucp_am.c [hunk 3]
```diff
@@ -470,43 +479,86 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
...
-                                  &reply_hdr, sizeof(reply_hdr),
-                                  ucp_proto_am_zcopy_req_complete);
+static ucs_status_t ucp_am_zcopy_single_reply(uct_pending_req_t *self)
+{
+    ucp_request_t *req            = ucs_container_of(self, ucp_request_t,
+                                                     send.uct);
+    unsigned user_hdr_length      = req->send.msg_proto.am.header_length;
+    size_t hdr_length             = sizeof(ucp_am_reply_hdr_t) +
+                                    user_hdr_length;
+    ucp_am_reply_hdr_t *reply_hdr = ucs_alloca(hdr_length);
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:522]: can user hdr be very big? what limits the stack allocation?

File: src/ucp/core/ucp_am.c [hunk 4]
```diff
@@ -470,43 +479,86 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
...
-                                 &first_hdr, sizeof(first_hdr), &mid_hdr,
-                                 sizeof(mid_hdr), ucp_proto_am_zcopy_req_complete,
-                                 1);
+    status = ucp_do_am_zcopy_multi(self, UCP_AM_ID_FIRST, UCP_AM_ID_MIDDLE,
+                                   first_hdr, hdr_length, NULL, sizeof(mid_hdr),
+                                   ucp_proto_am_zcopy_req_complete, 1);
+    if (status != UCS_ERR_NO_RESOURCE) {
+        ucs_assert(UCS_STATUS_IS_ERR(status) ||
+                   (req->send.state.dt.offset != 0));
+        req->send.length -= req->send.msg_proto.am.header_length;
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:558]: why in zcopy single we update length before and rever if failed, and in zcopy multi update length after?

File: src/ucp/core/ucp_am.c [hunk 5]
```diff
@@ -921,7 +991,7 @@ static ucs_status_t ucp_am_long_first_handler(void *am_arg, void *am_data,
         ucs_queue_del_iter(&ep_ext->am.mid_rdesc_q, iter);
         ucp_am_copy_data_fragment(first_rdesc, mid_hdr + 1,
                                   mid_rdesc->length - sizeof(*mid_hdr),
-                                  mid_hdr->offset + sizeof(*first_hdr));
+                                  mid_hdr->offset + first_rdesc->payload_offset);
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:994]: unrelated fix?

File: src/ucp/proto/proto_am.inl [hunk 6]
```diff
@@ -296,7 +296,8 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
         req->send.lane = ucp_send_request_get_am_bw_lane(req);
         ucp_send_request_add_reg_lane(req, req->send.lane);
     } else {
-        req->send.lane = ucp_ep_get_am_lane(ep);
+        req->send.lane         = ucp_ep_get_am_lane(ep);
+        req->send.pending_lane = req->send.lane;
```
Comment (yosefe) [inline src/ucp/proto/proto_am.inl:300]: why needed? it's done by ucp_request_pending_add()

File: src/ucp/core/ucp_am.c (hunk 3 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:522]: thought about it..
With zcopy, user header is limited by maximal UCT AM header (which may be quite large for certain transports)
do you think better use calloc starting from certain size?

File: src/ucp/core/ucp_am.c (hunk 5 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:994]: yes, but zcopy tests fail without it (when first fragment arrives later than second)

File: src/ucp/core/ucp_am.c (hunk 4 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:558]: will align with single zcopy

File: src/ucp/proto/proto_am.inl (hunk 6 above)
Comment (brminich) [reply src/ucp/proto/proto_am.inl:300]: To avoid returning `UCS_OK` for the first fragment. This way I can assume that UCS_OK will not be returned for the first fragment if there is no resources on AM lane

File: src/ucp/core/ucp_am.c (hunk 3 above)
Comment (yosefe) [reply src/ucp/core/ucp_am.c:522]: IMO we should have zcopy for user header as well, by using bounce buffer mpool

File: src/ucp/proto/proto_am.inl (hunk 6 above)
Comment (yosefe) [reply src/ucp/proto/proto_am.inl:300]: why it's needed for active messages? how is it different than eager zcopy for example?

Comment (brminich) [reply src/ucp/proto/proto_am.inl:300]: because if no resources returned  we will have to copy user header again and subtract its length from the req->send.length.
so we have to distinguish no_resource and all other cases to restore req->send.length

Comment (yosefe) [reply src/ucp/proto/proto_am.inl:300]: can we update send.length only after we are sure it was sent?

Comment (brminich) [reply src/ucp/proto/proto_am.inl:300]: i think we can, but it will be different from handling in single_zcopy case.
btw, why not initialize  pending_lane? We do not need to return UCS_OK if send was failed because of no_resources and req->send.pending_lane is null lane

Comment (yosefe) [reply src/ucp/proto/proto_am.inl:300]: it's not consistent with other usages of pending lane so makes the flow harder to understand. let's keep setting pending_lane in single place, as it is today.

Comment (brminich) [reply src/ucp/proto/proto_am.inl:300]: i did not get the point, but ok

File: src/ucp/core/ucp_am.c [hunk 7]
```diff
@@ -295,10 +318,12 @@ ucp_am_bcopy_pack_data(void *buffer, ucp_request_t *req, size_t length)
 {
     unsigned user_header_length = req->send.msg_proto.am.header_length;
 
-    ucs_assert((user_header_length == 0) || (length > user_header_length));
-    ucp_am_pack_user_header(buffer, req);
+    if (user_header_length != 0) {
+        ucs_assert(length > user_header_length);
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:326]: why > and not >= ? add comment

File: src/ucp/core/ucp_am.c [hunk 8]
```diff
@@ -479,86 +504,108 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
...
+        ucs_mpool_put_inline(req->send.msg_proto.am.reg_buf);
+        req->send.msg_proto.am.reg_buf = NULL;
+    }
+}
+
+void ucp_am_zcopy_completion(uct_completion_t *self, ucs_status_t status)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
+    if (req->send.state.dt.offset == req->send.length) {
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:529]: space line before

File: src/ucp/core/ucp_am.c [hunk 9]
```diff
@@ -479,86 +504,108 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
...
+    } else if (status != UCS_OK) {
+        ucs_assert(req->send.state.uct_comp.count == 0);
+        ucs_assert(status != UCS_INPROGRESS);
+
+        /* NOTE: the request is in pending queue if data was not completely sent,
+         *       just dereg the buffer here and complete request on purge
+         *       pending later.
+         */
+        ucp_request_send_buffer_dereg(req);
+        req->send.state.uct_comp.func = NULL;
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:540]: why needed?

File: src/ucp/proto/proto_am.inl [hunk 10]
```diff
@@ -222,28 +222,66 @@ void ucp_dt_iov_copy_uct(ucp_context_h context, uct_iov_t *iov, size_t *iovcnt,
...
+        iov[0].count  = 1;
+
+        ucp_dt_iov_copy_uct(ep->worker->context, iov + 1, &iovcnt, max_iov - 1,
+                            state, req->send.buffer, req->send.datatype,
+                            max_length - user_hdr_size, md_idx, NULL);
+        ++iovcnt;
+    } else {
+        ucp_dt_iov_copy_uct(ep->worker->context, iov, &iovcnt, max_iov,
+                            state, req->send.buffer, req->send.datatype,
+                            max_length, md_idx, NULL);
```
Comment (yosefe) [inline src/ucp/proto/proto_am.inl:253]: can we remove some code duplication by doing ++iov and --max_iov before calling ucp_dt_iov_copy_uct()?

File: src/ucp/core/ucp_am.c [hunk 11]
```diff
@@ -469,43 +509,108 @@ static ucs_status_t ucp_am_bcopy_multi(uct_pending_req_t *self)
...
+        if (req->send.msg_proto.am.reg_desc != NULL) {
+            ucs_mpool_put_inline(req->send.msg_proto.am.reg_desc);
+        }
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:543]: duplication of ucp_am_zcopy_req_complete()
why are we not completing user request here?
what prevents this request from generating additional fragments and using the (already released) memory registration?

File: src/ucp/core/ucp_am.c [hunk 12]
```diff
@@ -214,6 +214,11 @@ static UCS_F_ALWAYS_INLINE int ucp_am_recv_check_id(ucp_worker_h worker,
     return 1;
 }
 
+static UCS_F_ALWAYS_INLINE size_t
+ucp_am_send_req_total_size(ucp_request_t *req) {
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:218]: code style

File: src/ucp/core/ucp_am.c [hunk 13]
```diff
@@ -214,6 +214,11 @@ static UCS_F_ALWAYS_INLINE int ucp_am_recv_check_id(ucp_worker_h worker,
...
+    return req->send.length + req->send.msg_proto.am.header_length;
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:219]: maybe cache this value

File: src/ucp/proto/proto_am.inl (hunk 10 above)
Comment (brminich) [reply src/ucp/proto/proto_am.inl:253]: we can, but we will have to make a couple of unnecessary ops for empty header case. Imo, it looks ok - it just a single function invocation with different params.

File: src/ucp/core/ucp_am.c (hunk 7 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:326]: will add test and fix

File: src/ucp/core/ucp_am.c (hunk 11 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:543]: copied this from tag eager code, will double check

File: src/ucp/core/ucp_am.c (hunk 13 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:219]: request is full in the corresponding enum

File: src/ucp/proto/proto_am.inl (hunk 10 above)
Comment (yosefe) [reply src/ucp/proto/proto_am.inl:253]: this function is inlined and if we invoke it 2 times it will be more code

File: src/ucp/core/ucp_am.c (hunk 11 above)
Comment (yosefe) [reply src/ucp/core/ucp_am.c:543]: maybe need to fix tag eager as wel..

File: src/ucp/proto/proto_am.inl (hunk 10 above)
Comment (brminich) [reply src/ucp/proto/proto_am.inl:253]: ok

**Comment by brminich** [comment]
@yosefe, fixed
(had to force push due to wrong commit name, but it did not ruin review)

File: test/gtest/ucp/ucp_datatype.cc [hunk 14]
```diff
@@ -31,6 +31,11 @@ data_type_desc_t::make(ucp_datatype_t datatype, const void *buf, size_t length,
     m_dt = datatype;
     memset(m_iov, 0, sizeof(m_iov));
 
+    if (length == 0) {
+        m_is_empty = true;
+        m_count    = 0;
+    }
```
Comment (yosefe) [inline test/gtest/ucp/ucp_datatype.cc:37]: why need special case for empty? it can be contig

Comment (brminich) [reply test/gtest/ucp/ucp_datatype.cc:37]: to preserve validation interface in ucp_datatype class
(currently it assumes empty is non valid dt desc)

File: test/gtest/ucp/ucp_datatype.h [hunk 15]
```diff
@@ -88,9 +89,10 @@ class data_type_desc_t {
     };
 
     bool is_valid() const {
-        return (m_buf != NULL) && (m_count != 0) &&
-               (UCP_DT_IS_IOV(m_dt) ? (m_count <= m_iov_cnt_limit) :
-               (UCP_DT_IS_CONTIG(m_dt) || UCP_DT_IS_GENERIC(m_dt)));
+        return (m_is_empty && (m_count == 0)) ||
+               ((m_buf != NULL) && (m_count != 0) &&
+                (UCP_DT_IS_IOV(m_dt) ? (m_count <= m_iov_cnt_limit) :
```
Comment (yosefe) [inline test/gtest/ucp/ucp_datatype.h:94]: pls remove is_empty, check that count==0||m_buf!=NULL

Comment (yosefe) [reply test/gtest/ucp/ucp_datatype.h:94]: and check that m_dt is contig/iov/generic

Comment (brminich) [reply test/gtest/ucp/ucp_datatype.h:94]: this check is already here. We do not have invalid datatype, so can't use it as an indication of non-initialized desc.
That's why I added new boolean member

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
